### PR TITLE
Module 4

### PR DESCRIPTION
**Pull request by MiguelAlvarezHernanz**
**This branch includes my solution for Module_4. New files added:**
- A notebook file (`exploration_phase.ipynb`) with my approach and different attempts. (Already converted to `.md`)
- `push_model.py` contains a class which instantiates a calibrated Gradient Boosting Classifier.
- `module_4_fit_&_predict.py`, which contains a compilation of used functions, including both `handler_fit` and `handler_predict` for inference.

### CODE REVIEW

File: src/module_4/exploration_phase.md [hunk 1]
```diff
@@ -0,0 +1,293 @@
...
+    'min_child_weight': [1, 3, 5]  # Minimum sum of weights in a child node
+}
+
+xgb = XGBClassifier(eval_metric='logloss', random_state=42)
+
+random_search = RandomizedSearchCV(
+    xgb, 
+    param_distributions=param_grid, 
+    n_iter=10,
+    cv=2,
```
Comment (RafaZ7) [inline src/module_4/exploration_phase.md:242]: `cv=2` es un número un poco bajo, mejor intenta algo `>=3`. Cuanto más splits, más tiempo de entrenamiento, pero más fiables serán tus métricas.

File: src/module_4/exploration_phase.md [hunk 2]
```diff
@@ -0,0 +1,293 @@
...
+    
+![png](exploration_phase_files/exploration_phase_15_0.png)
+    
+
+
+
+```python
+from sklearn.model_selection import RandomizedSearchCV
+
+param_grid = {
```
Comment (RafaZ7) [inline src/module_4/exploration_phase.md:226]: Cuidado con usar `param_grids` tan extensos y con tan pocas iteraciones. Si te fijas tienes ~5000 combinaciones, pero solo estás testeando 10. Además los rangos para algunos parámetros son bastantes amplios (`max_depth` va de 3 a 10) por lo que al final va a ser casi un tema de suerte encontrar la mejor combinación.

Te recomiendo que para las primeras iteraciones pruebes con menos parámetros y más acotados, para empezar a tener cierta sensibilidad de qué valores pueden resultas óptimos en tu problema. Luego puedes hacer una iteración final con un mayor número de combinaciones y un número alto de iteraciones.

File: src/module_4/exploration_phase.md [hunk 3]
```diff
@@ -0,0 +1,293 @@
...
+    scoring='roc_auc', 
+    n_jobs=4, 
+    verbose=2,
+    random_state=42
+)
+
+random_search.fit(X_train, y_train)
+
+best_xgb = random_search.best_estimator_
+print(f"Best Parameters: {random_search.best_params_}")
```
Comment (RafaZ7) [inline src/module_4/exploration_phase.md:252]: Aquí estaría bien que generaras un dataframe con la información del cross validation, como qué métricas en train y validación tienes para cada combinación de parámetros. De esta forma puedes ver qué diferencias hay entre las distintas combinaciones, si la diferencia entre train y validación en muy alta (y te puede estar indicando que tu modelo está overfitteando mucho), etc.

Con el approach actual estás entrenando bastante a ciegas, porque tienes muchísimas combinaciones de parámetros, de las cuáles solo testeas 10 y te quedas con el mejor valor, pero sin tener ninguna sensibilidad de las métricas.

File: src/module_4/module_4_fit_&_predict.py [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+    
+    classifier_parametrisation = event.get("classifier_parametrisation", {})  # Diccionario de configuración del clasificador
+
+    calibration_parametrisation = event.get("calibration_parametrisation", {})  # Diccionario con parámetros de calibración
+    
+    prediction_threshold = event.get("prediction_threshold", 0.5)  # Valor por defecto si no está en el evento
+    
+    return model_folder_path, classifier_parametrisation, calibration_parametrisation, prediction_threshold
+
+def classifier_type(classifier_parametrisation: dict) -> str:
```
Comment (RafaZ7) [inline src/module_4/module_4_fit_&_predict.py:51]: Sé consistente con los espacios que dejas entre funciones. Siempre hay que dejar dos en python. Puedes usar `black` para que te formatee automáticamente los scripts.

File: src/module_4/module_4_fit_&_predict.py [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+def handler_fit(event: dict, _) -> dict[str, any]:
+    (
+        model_folder_path,
+        classifier_parametrisation,
+        calibration_parametrisation,
+        prediction_threshold,
+    ) = _extract_model_parameters(event)
+
+    df = load_training_feature_frame()
+    model_name = classifier_type(event)
```
Comment (RafaZ7) [inline src/module_4/module_4_fit_&_predict.py:80]: Puedes sacar el nombre de un modelo de forma genérica de esta forma (funciona para cualquier instancia de cualquier clase en python):
```python
model = XGBoostClassifier()
model_name = model.__class__.__name__
```

File: src/module_4/module_4_fit_&_predict.py [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+
+    predictions = clf.predict(data_to_predict)
+
+    probabilities = clf.predict_proba(data_to_predict)[:, 1]
+    predictions = (probabilities >= event.get("prediction_threshold")).astype(int)
+
+    return {
+        "statusCode": "200",
+        "body": {"prediction": predictions.to_dict()}
+    }
```
Comment (RafaZ7) [inline src/module_4/module_4_fit_&_predict.py:130]: Aquí me hubiera gustado ver algún ejemplo de cómo llamar a estos `handlers`.

File: src/module_4/module_4_fit_&_predict.py [hunk 7]
```diff
@@ -0,0 +1,130 @@
...
+    try:
+        df = pd.read_csv(file_path) 
+        return df
+    except FileNotFoundError:
+        raise Exception(f"Error: El archivo '{file_path}' no fue encontrado.")
+    except Exception as e:
+        raise Exception(f"Error al cargar los datos: {str(e)}")
+
+
+def _extract_model_parameters(event: dict) -> tuple[str, dict, dict, float]:
```
Comment (RafaZ7) [inline src/module_4/module_4_fit_&_predict.py:30]: Siempre que uses `dict` o `list` en el typing, intenta también definir cuáles qué tipos de elementos contienen.

Para este tipo de elementos como el event, que tienen una estructura compleja, podrías usar librerías como `pydantic` para validar en runtime el contenido. Tendrías algo así

```python
from pydantic import BaseModel, Field
from typing import Dict, Optional, Union, Any

class ModelEvent(BaseModel):
    """
    Pydantic model for ML model event parameters.
    """
    model_folder_path: str = Field(
        default="default_path/", 
        description="Path to the folder containing the model"
    )
    
    classifier_parametrisation: Dict[str, Any] = Field(
        default_factory=dict,
        description="Dictionary with classifier configuration parameters"
    )
    
    calibration_parametrisation: Dict[str, Any] = Field(
        default_factory=dict,
        description="Dictionary with model calibration parameters"
    )
    
    prediction_threshold: float = Field(
        default=0.5,
        description="Threshold for making positive predictions",
        ge=0.0,
        le=1.0
    )
```

Para usarlo:
```python
def process_event(event: Dict) -> tuple:
    """
    Extract model parameters from the received event.
    
    Args:
        event (Dict): Dictionary with input data
        
    Returns:
        Tuple[str, Dict, Dict, float]: Contains model path, classifier parameters, 
                                       calibration parameters, and prediction threshold
    """
    # Parse and validate the event with Pydantic
    model_event = ModelEvent(**event)
    
    # Access validated parameters
    return (
        model_event.model_folder_path,
        model_event.classifier_parametrisation,
        model_event.calibration_parametrisation,
        model_event.prediction_threshold
    )
```

**Review by RafaZ7 — APPROVED** [review]
Como siempre, puedes mergear cuando quieras!

Buen trabajo en general, el mayor punto a mejorar es la estrategia de gridsearch que has usado para el xgboost. Intenta empezar con rangos más pequeños para ir desarrollando la intuición de cómo se comporta cada parámetro (¿sube o baja el overfitting? ¿mejor o empeora las métricas en validación?).
